### PR TITLE
Performance Benchmarking

### PR DESCRIPTION
**Pull request by johanfylling**
Adding performance benchmark functionality

### CODE REVIEW

File: perf/jarl-perf.csv [hunk 1]
```diff
@@ -0,0 +1,4 @@
+label,baseline,perf/a,perf/b
+e3e755d17d,0.5786034939005645,4.833100232853919,7.443960715380243
+c0251ea41f,1.4706713179398083,5.908276495889907,9.236181934405773
+156807007b,0.40935575474398594,5.577915364549673,8.82058527772661
```
Comment (johanfylling) [inline perf/jarl-perf.csv:4]: We'll replace the contents of these CSV files once we can generate that using GitHub actions

File: .github/workflows/benchmark.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    - uses: actions/checkout@v3
+    - name: Update Jarl CSV
+      run: lein perf-jarl
+    - name: Create Pull Request
+      uses: peter-evans/create-pull-request@v4
+      commit-message: Updating performance metrics
+      signoff: true
+      branch: bench
+      delete-branch: true
+      branch-suffix: short-commit-hash
```
Comment (johanfylling) [inline .github/workflows/benchmark.yaml:19]: Doesn't seem like I can trigger this manually unless it's on `main` ..

File: src/test/resources/rego/perf/policy.rego [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package perf
+
+a := 42
+
+b := 1 + 2
```
Comment (johanfylling) [inline src/test/resources/rego/perf/policy.rego:5]: We should figure out some fun things to test here.

File: .github/workflows/benchmark.yaml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-name: Benchmark
+name: Benchmark Jarl
```
Comment (anderseknert) [inline .github/workflows/benchmark.yaml:1]: Will this only run on git push --tags? Or on any push?

File: project.clj [hunk 5]
```diff
@@ -13,22 +13,26 @@
   :aot [jarl.core jarl.parser jarl.api]
   :direct-linking true
   :aliases {"clj-kondo" ["with-profile" "+clj-kondo" "clj-kondo" "--debug" "--lint" "src"]
-            "eastwood" ["with-profile" "+eastwood" "eastwood"]
-            "kibit" ["with-profile" "+kibit" "kibit"]}
+            "eastwood"  ["with-profile" "+eastwood" "eastwood"]
+            "kibit"     ["with-profile" "+kibit" "kibit"]
+            "perf-opa"  ["with-profile" "+test" "run" "-m" "test.benchmark" "opa"]
+            "perf-jarl" ["with-profile" "+test" "run" "-m" "test.benchmark" "jarl"]}
```
Comment (anderseknert) [inline project.clj:19]: Nice

File: src/main/clojure/jarl/eval.clj [hunk 6]
```diff
@@ -477,3 +477,13 @@
                   (log/debugf "function <%s> returned undefined value" name)
                   {}))))
             (throw e)))))
+
+(defn find-plan [info entry-point]
```
Comment (anderseknert) [inline src/main/clojure/jarl/eval.clj:481]: Did we not have this somewhere before?

File: src/test/clojure/test/benchmark.clj [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+     :data       (conj (:data csv) entry)}))
+
+(defn git-revision []
+  (let [result (shell/sh "git" "rev-parse" "--short=10" "HEAD")
+        err (:err result)]
+    (when (not (string/blank? err))
+      (throw (ex-info "Failed to get OPA version" {:err err})))
+    (string/trim-newline (:out result))))
+
+(defn git-tag []
```
Comment (anderseknert) [inline src/test/clojure/test/benchmark.clj:103]: Lots of things going on in this namespace, from benchmarking, CSV handling, git and running shell commands.. I suppose that's fine as long as we don't need any of that elsewhere, in which case we should move some shared stuff.

File: src/test/resources/rego/perf/policy.rego (hunk 3 above)
Comment (anderseknert) [reply src/test/resources/rego/perf/policy.rego:5]: `numbers.range` is a favorite :)

**Review by anderseknert — APPROVED** [review]
Very nice! Looking forward to a demo :)

File: .github/workflows/benchmark.yaml (hunk 4 above)
Comment (johanfylling) [reply .github/workflows/benchmark.yaml:1]: Tags only. And when manually triggered.

File: src/main/clojure/jarl/eval.clj (hunk 6 above)
Comment (johanfylling) [reply src/main/clojure/jarl/eval.clj:481]: Funnily enough, no.

File: src/test/clojure/test/benchmark.clj (hunk 7 above)
Comment (johanfylling) [reply src/test/clojure/test/benchmark.clj]: Moved out git and csv stuff to their own namespaces.

**Pull request merged**